File: OrderManagement.py
class Product {
    String productId;
    String category;
    int quantity;
    double unitPrice;
    boolean inStock;

    Product(String productId, String category, int quantity,
            double unitPrice, boolean inStock) {

        this.productId = productId;
        this.category = category;
        this.quantity = quantity;
        this.unitPrice = unitPrice;
        this.inStock = inStock;
    }
}

public class OrderManagement {

    // Check whether product ID is valid
    static boolean isValidProduct(String productId) {
        return productId != null &&
               productId.matches("P[0-9]+");
    }

    // Category-specific discount
    static double calculateDiscount(String category, double amount) {

        switch(category.toLowerCase()) {

            case "electronics":
                return amount * 0.10;

            case "clothing":
                return amount * 0.15;

            case "grocery":
                return amount * 0.05;

            default:
                return 0;
        }
    }

    // GST calculation
    static double calculateGST(String category, double amount) {

        switch(category.toLowerCase()) {

            case "electronics":
                return amount * 0.18;

            case "clothing":
                return amount * 0.12;

            case "grocery":
                return amount * 0.05;

            default:
                return amount * 0.10;
        }
    }

    // Coupon discount
    // SAVE10 = 10%, maximum ₹500
    static double couponDiscount(String coupon, double amount) {

        if(coupon != null && coupon.equalsIgnoreCase("SAVE10")) {
            return Math.min(amount * 0.10, 500);
        }

        return 0;
    }

    // Calculate complete order
    static double calculateOrder(Product[] products, String coupon) {

        double subtotal = 0;
        double categoryDiscount = 0;
        double gst = 0;

        for(Product p : products) {

            if(!isValidProduct(p.productId)) {
                System.out.println("Invalid Product: " + p.productId);
                return -1;
            }

            if(p.quantity <= 0) {
                System.out.println("Invalid Quantity for " + p.productId);
                return -1;
            }

            if(p.unitPrice <= 0) {
                System.out.println("Invalid Price for " + p.productId);
                return -1;
            }

            if(!p.inStock) {
                System.out.println("Product Out of Stock: " + p.productId);
                return -1;
            }

            double amount = p.quantity * p.unitPrice;

            subtotal += amount;

            double discount = calculateDiscount(
                p.category, amount
            );

            categoryDiscount += discount;

            double taxableAmount = amount - discount;

            gst += calculateGST(
                p.category, taxableAmount
            );
        }

        // Coupon validation
        double couponDisc = couponDiscount(coupon, subtotal);

        if(coupon != null &&
           !coupon.equalsIgnoreCase("SAVE10") &&
           !coupon.equalsIgnoreCase("NONE")) {

            System.out.println("Invalid Coupon Code");
            return -1;
        }

        // Bulk discount
        int totalQuantity = 0;

        for(Product p : products) {
            totalQuantity += p.quantity;
        }

        double bulkDiscount = 0;

        if(totalQuantity >= 10) {
            bulkDiscount = subtotal * 0.05;
        }

        // Free shipping for subtotal >= ₹5000
        double shipping;

        if(subtotal >= 5000) {
            shipping = 0;
        } else {
            shipping = 100;
        }

        double finalAmount =
            subtotal
            - categoryDiscount
            - couponDisc
            - bulkDiscount
            + gst
            + shipping;

        return finalAmount;
    }

    // Display order summary
    static void displayOrder(Product[] products, String coupon) {

        double subtotal = 0;
        double categoryDiscount = 0;
        double gst = 0;

        int totalQuantity = 0;

        for(Product p : products) {

            double amount = p.quantity * p.unitPrice;

            subtotal += amount;

            categoryDiscount +=
                calculateDiscount(p.category, amount);

            gst += calculateGST(
                p.category,
                amount - calculateDiscount(p.category, amount)
            );

            totalQuantity += p.quantity;
        }

        double couponDisc = couponDiscount(coupon, subtotal);

        double bulkDiscount = 0;

        if(totalQuantity >= 10) {
            bulkDiscount = subtotal * 0.05;
        }

        double shipping = (subtotal >= 5000) ? 0 : 100;

        double finalAmount =
            subtotal
            - categoryDiscount
            - couponDisc
            - bulkDiscount
            + gst
            + shipping;

        System.out.println("\n------- ORDER SUMMARY -------");
        System.out.println("Subtotal            : " + subtotal);
        System.out.println("Category Discount   : " + categoryDiscount);
        System.out.println("Coupon Discount     : " + couponDisc);
        System.out.println("Bulk Discount       : " + bulkDiscount);
        System.out.println("GST                 : " + gst);
        System.out.println("Shipping Charge     : " + shipping);
        System.out.println("-----------------------------");
        System.out.println("Final Amount        : " + finalAmount);
    }

    public static void main(String[] args) {

        // Normal development test order
        Product[] products = {

            new Product(
                "P101",
                "electronics",
                2,
                2000,
                true
            ),

            new Product(
                "P102",
                "clothing",
                3,
                1000,
                true
            ),

            new Product(
                "P103",
                "grocery",
                5,
                200,
                true
            )
        };

        String coupon = "SAVE10";

        double result = calculateOrder(products, coupon);

        if(result != -1) {
            displayOrder(products, coupon);
        }
    }
}

File: OrderManagement.java
import java.util.*;

public class OrderManagement {

    static class Product {
        String id;
        String category;
        double price;
        int stock;

        Product(String id, String category, double price, int stock) {
            this.id = id;
            this.category = category;
            this.price = price;
            this.stock = stock;
        }
    }

    static Map<String, Product> products = new HashMap<>();

    static {
        products.put("P101", new Product("P101", "Electronics", 1000, 20));
        products.put("P102", new Product("P102", "Clothing", 800, 30));
        products.put("P103", new Product("P103", "Grocery", 500, 50));
        products.put("P104", new Product("P104", "Books", 400, 25));
        products.put("P105", new Product("P105", "Electronics", 2000, 5));
    }

    public static double calculateOrder(String[] productIds, int[] quantities,
                                        String coupon) {

        double subtotal = 0;
        int totalQuantity = 0;

        for (int i = 0; i < productIds.length; i++) {

            if (!products.containsKey(productIds[i])) {
                throw new IllegalArgumentException(
                        "Invalid product: " + productIds[i]);
            }

            if (quantities[i] <= 0) {
                throw new IllegalArgumentException(
                        "Quantity must be greater than zero");
            }

            Product p = products.get(productIds[i]);

            if (quantities[i] > p.stock) {
                throw new IllegalArgumentException(
                        "Out of stock: " + productIds[i]);
            }

            subtotal += p.price * quantities[i];
            totalQuantity += quantities[i];
        }

        double categoryDiscount = 0;

        for (int i = 0; i < productIds.length; i++) {

            Product p = products.get(productIds[i]);
            double amount = p.price * quantities[i];

            if (p.category.equalsIgnoreCase("Electronics")) {
                categoryDiscount += amount * 0.10;
            } else if (p.category.equalsIgnoreCase("Clothing")) {
                categoryDiscount += amount * 0.15;
            } else if (p.category.equalsIgnoreCase("Books")) {
                categoryDiscount += amount * 0.05;
            }
        }

        double bulkDiscount = 0;

        if (totalQuantity >= 10) {
            bulkDiscount = subtotal * 0.05;
        }

        double couponDiscount = 0;

        if (coupon.equalsIgnoreCase("SAVE10")) {
            couponDiscount = subtotal * 0.10;
        } else if (coupon.equalsIgnoreCase("SAVE20")) {
            couponDiscount = subtotal * 0.20;
        } else if (!coupon.equalsIgnoreCase("NONE")) {
            throw new IllegalArgumentException("Invalid coupon code");
        }

        double totalDiscount =
                categoryDiscount + bulkDiscount + couponDiscount;

        double maximumDiscount = subtotal * 0.30;

        if (totalDiscount > maximumDiscount) {
            totalDiscount = maximumDiscount;
        }

        double amountAfterDiscount = subtotal - totalDiscount;

        double gst = amountAfterDiscount * 0.18;

        double shipping;

        if (amountAfterDiscount >= 5000) {
            shipping = 0;
        } else {
            shipping = 100;
        }

        double finalAmount =
                amountAfterDiscount + gst + shipping;

        System.out.println("----------------------------------");
        System.out.println("E-COMMERCE ORDER");
        System.out.println("----------------------------------");
        System.out.println("Products         : " +
                Arrays.toString(productIds));
        System.out.println("Quantities       : " +
                Arrays.toString(quantities));
        System.out.println("Coupon           : " + coupon);
        System.out.println("Subtotal         : " + subtotal);
        System.out.println("Category Discount: " + categoryDiscount);
        System.out.println("Bulk Discount    : " + bulkDiscount);
        System.out.println("Coupon Discount  : " + couponDiscount);
        System.out.println("Total Discount   : " + totalDiscount);
        System.out.println("GST (18%)        : " + gst);
        System.out.println("Shipping Charge  : " + shipping);
        System.out.println("Final Amount     : " + finalAmount);
        System.out.println("----------------------------------");

        return finalAmount;
    }

    public static void main(String[] args) {

        String[] productIds = {"P101", "P102", "P104"};
        int[] quantities = {2, 3, 2};
        String coupon = "SAVE10";

        try {

            double amount = calculateOrder(
                    productIds,
                    quantities,
                    coupon);

            System.out.println("ORDER PROCESSED SUCCESSFULLY");
            System.out.println("Amount to Pay: " + amount);

        } catch (Exception e) {

            System.out.println("ORDER FAILED");
            System.out.println("Reason: " + e.getMessage());
        }
    }
}
